drivers/homeless_summary_report/app/controllers/homeless_summary_report/warehouse_reports/reports_controller.rb

```diff
@@ -0,0 +1,169 @@
+    before_action :set_variants, only: [:show, :details]
+    before_action :set_m1_fields, only: [:show, :details]
+    before_action :set_m2_fields, only: [:show]
+    before_action :set_m7_fields, only: [:show, :details]
+
+    def index
+      @reports = report_scope.ordered.
+        page(params[:page]).per(25)
+      @report = report_class.new(user_id: current_user.id)
+      # Make sure the form will work
+      filters
+    end
+
+    def show
+      @measures = {
+        'Measure 1': @m1_fields.keys,
+        'Measure 2': @m2_fields,
+        'Measure 7': @m7_fields,
+      }
+    end
+
+    def create
+      @report = report_class.new(
+        user_id: current_user.id,
+      )
+      @report.filter = @filter
+      @report.save
+      # TODO: Now we need to figure out how to define the job. I think it's by modifying the report class, probably ultimately the run_and_save! method. Yes.
+      ::WarehouseReports::GenericReportJob.perform_later(
+        user_id: current_user.id,
+        report_class: @report.class.name,
+        report_id: @report.id,
+      )
+      # Make sure the form will work
+      filters
+      respond_with(@report, location: homeless_summary_report_warehouse_reports_reports_path)
+    end
+
+    def destroy
+      @report.destroy
+      respond_with(@report, location: homeless_summary_report_warehouse_reports_reports_path)
+    end
+
+    def details
+      @variant = details_params['variant'] || 'all_persons'
+      @cell = "spm_#{details_params['cell']}" || 'spm_m1a_es_sh_days'
+      @cell_name = @cell.humanize.split(' ')[1..].join(' ')
+      if details_params['cell'].start_with?('m1')
+        @measure = 'Measure 1'
+        @data_cells = @m1_fields.keys
+      elsif details_params['cell'].start_with?('m2')
+        @measure = 'Measure 2'
+        @data_cells = ['m2_reentry_days']
+      elsif details_params['cell'].start_with?('m7')
+        @measure = 'Measure 2'
+        @data_cells = @m7_fields
+      end
+      @detail_clients = @report.clients.send(@variant).send(@cell)
+      @spm_id = @detail_clients.first.send("spm_#{@variant}")
+    end
+
+    def details_params
+      params.permit(
+        :variant,
+        :cell,
+      ).delete_if do |key, value|
+        key == 'variant' && report_class.report_variants.keys.exclude?(value.to_sym)
+      end.delete_if do |key, value|
+        key == 'cell' && (report_class.spm_fields.keys + [
+          :m2_reentry_0_to_180_days,
+          :m2_reentry_181_to_365_days,
+          :m2_reentry_366_to_730_days,
+          :m7a1_c2,
+          :m7a1_c3,
+          :m7a1_c4,
+          :m7b1_c2,
+          :m7b1_c3,
+          :m7b2_c2,
+          :m7b2_c3,
+        ]).exclude?(value.to_sym)
+      end
+    end
+    helper_method :details_params
+
+    private def set_report
+      @report = report_class.find(params[:id].to_i)
+    end
+
+    private def set_variants
+      @variants = ::HomelessSummaryReport::Report.report_variants
+    end
```

Comment:
I probably wouldn't bother with this, just reference `report_class.report_variants`, or setup an instance method that memoizes that so you can call `@report.variants` instead of an extra instance variable.